pkg/boot/grub/grub.go

```diff
@@ -0,0 +1,405 @@
+	p := newParserWithSchemes(wd, s)
+	if err := p.appendFile(url); err != nil {
+		return nil, err
+	}
+	return p.config, nil
+}
+
+type parser struct {
+	config *Config
+	W      io.Writer
+
+	// parser internals.
+	scope    scope
+	numEntry int
+	curEntry string
+	curLabel string
+	wd       *url.URL
+	schemes  curl.Schemes
+}
+
+type scope uint8
+
+const (
+	scopeGlobal scope = iota
+	scopeEntry
+)
+
+// newParserWithSchemes returns a new grub parser using working directory `wd`
+// and schemes `s`.
+//
+// If a path encountered in a configuration file is relative instead of a full
+// URL, `wd` is used as the "working directory" of that relative path; the
+// resulting URL is roughly `wd.String()/path`.
+//
+// `s` is used to get files referred to by URLs.
+func newParserWithSchemes(wd *url.URL, s curl.Schemes) *parser {
+	return &parser{
+		config: &Config{
+			Entries: make(map[string]boot.OSImage),
+		},
+		scope:   scopeGlobal,
+		wd:      wd,
+		schemes: s,
+	}
+}
+
+func parseURL(surl string, wd *url.URL) (*url.URL, error) {
+	u, err := url.Parse(surl)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse URL %q: %v", surl, err)
+	}
+
+	if len(u.Scheme) == 0 {
+		u.Scheme = wd.Scheme
+
+		if len(u.Host) == 0 {
+			// If this is not there, it was likely just a path.
+			u.Host = wd.Host
+			u.Path = filepath.Join(wd.Path, filepath.Clean(u.Path))
+		}
+	}
+	return u, nil
+}
+
+// getFile parses `url` relative to the config's working directory and returns
+// an io.Reader for the requested url.
+//
+// If url is just a relative path and not a full URL, c.wd is used as the
+// "working directory" of that relative path; the resulting URL is roughly
+// path.Join(wd.String(), url).
+func (c *parser) getFile(url string) (io.ReaderAt, error) {
+	u, err := parseURL(url, c.wd)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.schemes.LazyFetch(u)
+}
+
+// appendFile parses the config file downloaded from `url` and adds it to `c`.
+func (c *parser) appendFile(url string) error {
+	r, err := c.getFile(url)
+	if err != nil {
+		return err
+	}
+	config, err := uio.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	if len(config) > 500 {
+		// Avoid flooding the console on real systems
+		// TODO: do we want to pass a verbose flag or a logger?
+		log.Printf("Got config file %s", r)
+	} else {
+		log.Printf("Got config file %s:\n%s\n", r, string(config))
+	}
+	return c.append(string(config))
+}
+
+func isWhitespace(b byte) bool {
+	return b == '\t' || b == '\n' || b == '\v' ||
+		b == '\f' || b == '\r' || b == ' '
+}
+
+type quote uint8
+
+const (
+	unquoted quote = iota
+	escape
+	singleQuote
+	doubleQuote
+	doubleQuoteEscape
+	comment
+)
+
+// Fields splits a grub line and unquote it's components according to
+// https://www.gnu.org/software/grub/manual/grub/grub.html#Quoting
+// except that the escaping of newline is not supported
+func fields(s string) []string {
```

Comment:
I haven't looked at this too closely but on the surface this seems to be similar to https://github.com/u-root/u-root/blob/master/pkg/pogosh/lexer.go#L92 
Is there a way we could reuse some code between them?

Reply:
could we get this in and then attack that change?

Reply:
I asked Ryan the same thing before, when pogosh got committed. I dislike the fact that we rush reviews only because we've been letting them sit around for a while. We should be faster reviewing stuff, and we should not just submit things because they happen to exist.

Ryan, however, did convince me that reusing that code is not that easy. I'm okay keeping this here for now.

Reply:
I dislike it, btw, because we say "can we do that later?" and then we never do.

Reply:
but that applies to reviews too, right? It's not that this is rushed -- it's been in use for a while. At some point the burden shifts to the reviewer to deliver on suggestions along the lines of "wait a while, we can do this better", because in all too many cases, we never get to that either. We sometimes accept something that is in use and may not fit the form a reviewer would like. This commit has been there for almost 5 months.

Reply:
Also when I started this, pogosh was not in, I asked wether using evlish whould do (same idea) but was discouraged.
Anyway yes the lexer of pogosh looks similar, the good thing is we can reuse the tests from grub when trying to refactor.